### PR TITLE
Switch `SwiftSelf<T>` position requirement to last

### LINKED ISSUE #107946: [swift interop] Cannot call instance methods with at least one parameter when using `SwiftSelf<T>`
Our current interop implementation only allows SwiftSelf<T> (representing the self parameter in Swift instance methods) to be specified as the first parameter in function signatures. This design choice leads to issues when interfacing with Swift instance methods that accept additional parameters.

Let's consider following example

```swift
@frozen public struct SampleStruct
{
    public var x1: Int
    public var x2: Int

    public func addWithExtraArgs(x3: Float, x4: Float) -> Float {
        return Float(x1 + x2) + x3 + x4
    }
}
```

`addWithExtraArgs` results in the following LLVM-IR:

```llvm
define protected swiftcc float @"output.SampleStruct.addWithExtraArgs(x3: Swift.Float, x4: Swift.Float) -> Swift.Float"(float %0, float %1, i64 %2, i64 %3) ...
```

`SampleStruct` is expected to be lowered into two `i64` which are passed as **last** parameters.
This would correspond to following csharp signature:
```c#
static extern void AddWithExtraArgs(float a, float b, SwiftSelf<SampleStruct> struct)
```

We should either allow `SwiftSelf<T>`  argument to be specified as the last parameter or we should reorder the parameters in runtime.

\cc: @kotlarmilos, @jakobbotsch, @amanasifkhalid, @matouskozak

### PR DESCRIPTION
**Pull request by jakobbotsch**
This is added as the last parameter by the Swift compiler, so to support instance calls taking parameters this needs to come last.

Fix #107946

### CODE REVIEW

**Comment by jakobbotsch** [comment]
@jkurdek @kotlarmilos any idea on the Mono crash?

**Comment by jkurdek** [comment]
I tried reproing the mono failure locally. Both on arm64 and x64. The tests passed without problems. There might be something less obvious going on. I will take another look next week.

File: src/tests/Interop/Swift/SwiftSelfContext/SwiftSelfContext.cs [hunk 1]
```diff
@@ -81,4 +81,26 @@ public unsafe static void TestSelfIsFrozenNonEnregisteredStruct()
...
+
+    [Fact]
+    public unsafe static void TestSelfIsFrozenEnregisteredStructWithExtraArgs()
+    {
+        float sum = SumFrozenEnregisteredStructWithExtraArgs(3f, 4f, new SwiftSelf<FrozenEnregisteredStruct>(new FrozenEnregisteredStruct { A = 10, B = 20 }));
+        Assert.Equal(37f, sum);
+    }
+
+    [Fact]
+    public unsafe static void TestSelfIsFrozenNonEnregisteredStructWithExtraArgs()
```
Comment (jkurdek) [inline src/tests/Interop/Swift/SwiftSelfContext/SwiftSelfContext.cs:101]: It seems that this is the test which fails on the CI. So far I have determined that the error also occurs when we use non-generic SwifSelf instead. The error seems to be only occurring on the runtime x64 macos runtime_tests pipeline.

Comment (jakobbotsch) [reply src/tests/Interop/Swift/SwiftSelfContext/SwiftSelfContext.cs:102]: I opened #108855 about this and disabled this test against that issue

**Review by jkurdek — APPROVED** [review]
LGTM! Thanks for opening mono tracking issue!

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by jakobbotsch** [comment]
/ba-g The chrome-DebuggerTests look like #108078, but were "dead lettered" so build analysis is not picking it up

**Pull request merged**